src/classes/PMT_Payment_TDTM.cls

```diff
@@ -201,4 +162,132 @@ public class PMT_Payment_TDTM extends TDTM_Runnable {
         }
     }
 
+    /*******************************************************************************************************
+    * @description          Retrieve a Id Map of Opportunties related to a Set of Ids
+    * @param oppIds         Set of Opportunity Ids to retrieve
+    * @return Map<Id, Opp>  This method returns a map of Id to opportunity
+    ********************************************************************************************************/
+    private Map<Id, Opportunity> getPaymentOpportunities(Set<Id> oppIds) {
+        String soql = 'SELECT Id, Amount, IsClosed, IsWon, StageName, ';
+        if (isCurrencyFieldAvailable) {
+            soql += 'CurrencyIsoCode, ';
+        }
+        soql += ' (SELECT id, npe01__Paid__c, npe01__Payment_Amount__c, npe01__Payment_Date__c, npe01__Written_Off__c ' +
+                ' FROM Opportunity.npe01__OppPayment__r)' +
+                ' FROM Opportunity WHERE Id IN :oppIds';
+
+        return new Map<Id, Opportunity>((List<Opportunity>)Database.query(soql));
+    }
+
+    /*******************************************************************************************************
+    * @description          For multicurrency orgs, validate the currencies and set the appropriate stages
+    *                       and currency fields
+    * @param newlist        The list of Payments from trigger new.
+    * @param oldlist        The list of Payments from trigger old.
+    * @param triggerAction  Which trigger event (BeforeInsert, AfterInsert, etc.).
+    * @return dmlWrapper    This method modifies dmlWrapper and returns it
+    ********************************************************************************************************/
+    private DmlWrapper handleMultiCurrencyPayments(
+        List<npe01__OppPayment__c> newlist, 
+        List<npe01__OppPayment__c> oldlist,
+        TDTM_Runnable.Action triggerAction, 
+        DmlWrapper dmlWrapper) {
+
+        if (dmlWrapper == null) {
+            dmlWrapper = new DmlWrapper();
+        }
+
+        Set<Id> oppIds = new Set<Id>();
+        for (npe01__OppPayment__c pmt : newlist) {
+            oppIds.add(pmt.npe01__Opportunity__c);
+        }
+
+        Map<Id, Opportunity> mapOfOpps = getPaymentOpportunities(oppIds);
+        
+        // Validation and default the CurrencyIsoCode for new records for Before Insert/update
+        if (isCurrencyFieldAvailable && (triggerAction == TDTM_Runnable.Action.BeforeInsert || triggerAction == TDTM_Runnable.Action.BeforeUpdate)) {
```

Comment:
I'm not sure if we should change this, but I've read that trigger validation code should go in the after trigger:

https://andyinthecloud.com/2015/04/19/considerations-placing-validation-code-in-an-apex-triggers/

In short, if you have multiple before triggers on the same object (client code, other package code) and since we can't control trigger order, there's a possibility that the validation that happens in this trigger could be circumvented in another subsequent trigger. By putting validation in an after trigger, the order doesn't matter since you can't edit the trigger records at that point. @force2b what do you think?

Reply:
@seanpat09 That's a good point.  I'm not sure about the original intent of the validations but it is something to consider.

Reply:
Good point @seanpat09. I'm hesitant to change the trigger behavior though because it might impact existing implementations. One thing that does come to mind is a scenario where the customers Before trigger changes the Paid and Written-Off fields both to checked, and if that trigger were to run _after_ our Before trigger it is feasible that our validation would not pick that up. 

The Order of Execution documentation states that Salesforce will re-execute most steps in the process in a recursive save (which includes Before & After triggers). My interpretation of this is that if an After trigger modifies a record (such as to check both boxes), the Before trigger will run again and ideally validation should pick up that change. I'd like to test that though getting Salesforce to execute triggers in a specific order is not all that easy.

Is there a way we can pull off a solid trigger execution order test to confirm behavior?

Reply:
@force2b @seanpat09 , so i was testing this - and there's what I found:

As we have mentioned it is possible that before triggers miss validating the record because of the order of execution.  If before trigger one validates and then before trigger two changes it - the record could be marked as both paid and written off.

If the after trigger does change it (recursively) then the before validation fires again - so that does catch it.  

I believe that for our NEW validation check (the one for written and paid off) we can use the after trigger (since it is new logic).  But this particular code here - actually does SET and Validation so we don't want to move it to an after trigger.  It is something to be aware of but I would say that we should probably not convert it at this point.

All the other code review changes should have been made.  Thanks

Reply:
Thanks for researching @patrick-yan-sf. Are you confirming that if the After trigger modifies the record, the Before trigger is executed again? If that's the case, is there value in moving the validations (both the written-off/paid & payment date) to an After trigger.

Reply:
@force2b - yes - what I did was create an after trigger on payment that updated the payment record (to have both written off and paid true).  The validation for that logic was on the before trigger - and you get the before trigger validation error.

@RadSFDeveloper @seanpat09  - do you have any thoughts on this?  I think in general you are correct - we can change the other date validation to be an after update and after insert as well.

Reply:
I agree, having the validation in the after trigger will prevent any customer before triggers from bypassing our validation.

For example, if the customer writes a before trigger that updates both written off/paid checkboxes to true (for whatever reason), and it runs following our before trigger, then the validation will be skipped. Having this validation in the after trigger ensures that this situation is prevented and also covers recursive updates in workflows/updates triggers, etc.

You're right in that this is something we should only do with new validations we create going forward so that we don't cause changes to existing behavior

Reply:
Thanks @seanpat09. We will live the existing behavior for now.

FYI - I've also just created the PR for the actual validation.  If someone has time to review these then maybe we can get it to QE tomorrow.  Thanks